polling_stations/apps/data_collection/data_types.py

```diff
@@ -164,6 +164,68 @@ def append(self, address):
                 pretty=True,
             )
 
+    def validate_postcodes(self, addressbase_data):
+        """
+        This method checks if a record has a valid postcode.
+        If it doesn't then it looks for uprn, and if it can match the uprn in
+        addressbase, and the postcodes look similar it will correct the postcode
+        for us.
+        Even if it doesn't match the uprn it saves us some legwork by having tried
+        various options.
+        """
+        postcode_validator = GBPostcodeField()
+
+        def invalid_postcode_warning(record, extras=""):
+            """Game over, chuck it out, explain why"""
+            self.logger.log_message(
+                logging.WARNING,
+                f"Discarding record with invalid postcode{extras}.\nRecord:\n%s\n",
+                variable=record,
+                pretty=True,
+            )
+            self.elements.remove(record)
+
+        def correct_postcode_from_uprn(record, addressbase_data):
+            if record["uprn"] not in addressbase_data:
+                invalid_postcode_warning(
+                    record, extras=" and record UPRN not in addressbase"
+                )
+
+            addressbase_record = addressbase_data[record["uprn"]]
+            if "postcode" in addressbase_record:
+                # At the moment this just suggests a correction from ABP if possible.
+                # It would be better to catch the obvious errors and just fix them.
+                # eg. '0'=>'O', 'I|L'=>'1', etc.
```

Comment:
How about using the same approach here that we use in `handle_invalid_uprns()`? Something like:

```py
if fuzzy_match:
    match_quality = fuzz.partial_ratio(
        record["address"].lower().replace(",", ""),
        addressbase_record["address"].lower().replace(",", "")
    )
    if match_quality >= match_threshold:
        # accept the correction from addressbase
    else:
        # bounce it back as a suggestion
```

That's probably going to do the job for us in most cases.